### DISCUSSION TITLE
Inquiry about using Exodus mesh format for initial grain structure

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#24328] [Q&A Modules: Phase field] [answered]
Dear MOOSE experts,

I am writing to seek your advice on how to use an exodus mesh format file as input to create an initial grain structure for phase-field simulation of grain growth. Specifically, I want to utilize the `ElementBlockIds` information in the mesh file to create the initial grain structure.
[out_test4_01.e.38.05.txt](https://github.com/idaholab/moose/files/11453129/out_test4_01.e.38.05.txt)

As you can see from the attached figure, the `ElementBlockIds` in the exodus file are labeled differently for each grain. I want to use this information to assign different grain IDs to each grain and create an initial grain structure.
![image](https://github.com/idaholab/moose/assets/56183775/e7125ac1-1725-4489-84c5-9542b0021fe8)

Could you please guide me on how to achieve this in MOOSE? Any suggestions or examples would be greatly appreciated.

Thank you very much for your time and help.

Best regards,
wei

**Comment by GiudGiud** [comment]
@laagesen @amjokisaari

**Comment by laagesen** [comment]
Sorry I don't have a good idea off the top of my head

**Reply by GiudGiud** [reply]
How is a grain defined in the code again? You keep track of the boundary with variables right?

**Reply by laagesen** [reply]
No each grain is represented with an order parameter, a number that is 1 inside a given grain and 0 outside, with a diffuse interface between 0 and 1 at interfaces. So in the absence of the GrainTracker, if we have 20 grains, we need 20 order parameters. The GrainTracker allows 1 order parameter to represent multiple grains to be represented by 1 OP by monitoring for overlaps and re-assigning grains if necessary. The GrainTracker does keep a unique ID for each grain so maybe this data could be mapped into the unique ID but I'm not sure if you can get the order parameter mapping directly from that.

**Reply by GiudGiud** [reply]
is this kept on a per-element basis?

**Reply by laagesen** [reply]
I think you can select whether the unique IDs are stored at nodes or elements but I should ask @permcody to confirm this

**Reply by PengWei97** [reply]
I believe can simplify the problem further. Suppose we have a mesh file (.e/.vtk) that already stores GrainID or ElementBlockIds for each node/element. Can we use this file as the initial microstructure for grain growth simulations with GrainTracker, instead of using built-in methods such as VT algorithm to establish the initial microstructure? Additionally, I believe there may be some insights we can gain from using EBSD as an input file for the simulation.

**Reply by GiudGiud** [reply]
If you can load them into variables, can the FeatureFloodCount do that?

@permcody on initializing grains from Exodus/EBSD

**Reply by permcody** [reply, marked as answer]
We do have an [ImageFunction](https://mooseframework.inl.gov/moose/source/functions/ImageFunction.html) object which is pretty close to what you need, but I don't think it's exactly what you'd need for a polycrystal simulation without a little additional development. I've wanted to create an image reader for starting polycrystal simulations for many years, but I just don't have the time. However, the [PolycrystalUserObjectBase](https://mooseframework.inl.gov/modules/phase_field/ICs/PolycrystalICs.html) was designed with exactly this use case in mind:

Essentially the user interface is to query each element in your domain and your job as the developer of a derived class is supply one or more active grains at that location. That's it! That object will do the proper assignment of those grains to order parameters and your simulation will run. The way I'd approach this task would be to investigate the capabilities of that ImageFunction object and see if you can get unique values for each color in your image, if you can, then you are pretty much done on creating an "ImageFunctioPolycrystalIC" class. 

Does this make sense? Take a look and ask me questions!
